test/cpp/microbenchmarks/callback_test_service.h

```diff
@@ -0,0 +1,114 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef TEST_CPP_MICROBENCHMARKS_CALLBACK_TEST_SERVICE_H
+#define TEST_CPP_MICROBENCHMARKS_CALLBACK_TEST_SERVICE_H
+
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/cpp/util/string_ref_helper.h"
+
+namespace grpc {
+namespace testing {
+
+const char* const kServerFinishAfterNReads = "server_finish_after_n_reads";
+const char* const kServerResponseStreamsToSend = "server_responses_to_send";
+
+class CallbackStreamingTestService
+    : public EchoTestService::ExperimentalCallbackService {
+ public:
+  CallbackStreamingTestService() {}
+  void Echo(ServerContext* context, const EchoRequest* request,
+            EchoResponse* response,
+            experimental::ServerCallbackRpcController* controller) override;
+
+  experimental::ServerBidiReactor<EchoRequest, EchoResponse>* BidiStream()
+      override;
+};
+
+class BidiClient
+    : public grpc::experimental::ClientBidiReactor<EchoRequest, EchoResponse> {
+ public:
+  BidiClient(EchoTestService::Stub* stub, EchoRequest* request,
+             EchoResponse* response, ClientContext* context,
+             int num_msgs_to_send)
+      : request_{request},
+        response_{response},
+        context_{context},
+        msgs_to_send_{num_msgs_to_send} {
+    stub->experimental_async()->BidiStream(context_, this);
+    MaybeWrite();
+    StartRead(response_);
+    StartCall();
+  }
+
+  void OnReadDone(bool ok) override {
+    if (ok && reads_complete_ < msgs_to_send_) {
```

Comment:
Nit: to be consistent with below, you can move the ok out and do:
if (!ok) {
  return;
}